### PR TITLE
0.1.0

### PR DESCRIPTION
**Pull request by xarvh**
* Add assertion for invalid schemas (string/number rather than function/array/object)
* Remove async stuff
* Rewrite tests in coffee
* Begin the readme with an actual example

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,119 +1,231 @@
...
+  var theDesiredOutput = {
+    id: 123,
+    name: 'HappyLand',
+    base: {
+      color: '#7FF',
+      time: 'Mon - Aug'
+    }
+  }
+
+  // enters Cartography
```
Comment (etaoins) [inline README.md]: What does `enters Cartography` mean here? I'd normally read that as "execution enters Cartography code" but this is a simple assignment.

File: README.md [hunk 2]
```diff
@@ -1,119 +1,231 @@
...
+    }
+  }
+
+  assert.deepEqual(cartography.map(theInput, theSchemaToTranslateOneIntoTheOther), theDesiredOutput)
 ```
-(All examples are in CoffeeScript, but Cartography is pure JavaScript.)
 
-The schema's structure resembles that of the output Object: it has the
+The schema's structure is the same as that of the output Object: it has the
 same attribute names and the same nested structure (if any), but each
```
Comment (etaoins) [inline README.md]: This isn't true if some of the mappers returned `undefined`, right? Not sure if that's worth clarifying

File: README.md [hunk 3]
```diff
@@ -1,119 +1,231 @@
...
-`id: from 'id', required`.
-
-A *filter* can be any function that accepts an argument as its input value
-and returns an output value.
-Filters can be used for checking the input validity by throwing a
-`CartographyError` if the input is not valid and returning it normally
-otherwise.
+schema Object or a custom function.
+
+**Lists of filters** are Array of filters.
```
Comment (etaoins) [inline README.md]: Array**s** of filters?

File: README.md [hunk 4]
```diff
@@ -1,119 +1,231 @@
...
+MAIN API
+--------
+
+### map(source, schema)
+Translates the `source` object according to the given `schema` and returns the result.
+Any field of the schema whose final value is `undefined` will not appear in the result.
+If no fields have defined values, `map` will return `undefined` rather than an empty Object.
+
+
+### CartographyError()
```
Comment (etaoins) [inline README.md]: `CartographyError(message)`?

File: README.md (hunk 1 above)
Comment (xarvh) [inline README.md]: I guess it was theatrical.
Would `// let's use Cartography` be more clear?
Shall I remove the comment entirely?

File: README.md (hunk 3 above)
Comment (xarvh) [inline README.md]: :+1:

File: README.md (hunk 4 above)
Comment (xarvh) [inline README.md]: Yessss! Good catch!

File: README.md [hunk 5]
```diff
@@ -1,119 +1,231 @@
...
-In addition to the usual synchronous filters and custom functions, `asyncMap`
-can use asynchronous filters and asynchronous custom functions provided they
-are wrapped by the `async` function decorator.
-The `asyncArray` filter should also be used in place of `array`.
+### filters.parseJSON
+Converts a JSON string into a JavaScript Object.
 
 
+### filters.isString
+Throws if the value is not a string (will reject String objects).
```
Comment (etaoins) [inline README.md]: This isn't clear to me. If it rejects `String` objects does it only accept string literals? Might be good to call that out explicitly for people that aren't as well versed in the joys of JavaScript

File: README.md [hunk 6]
```diff
@@ -1,119 +1,231 @@
...
+
+
+### filters.isNumber
+Throws if the value is not a number (will reject Number objects).
```
Comment (etaoins) [inline README.md]: Same here - does this mean it only accept string literals?

File: index.js [hunk 7]
```diff
@@ -20,124 +13,53 @@ var cartography = (function(){
...
+      var sourceAttributePath, filters, value;
 
-      var value = sourceAttribute.split('.').reduce((function(o, step) {return (o || {})[step];}), source);
-      return runFiltersOn(value, filters, function(message) {return addSeparator(sourceAttribute, message);});
-    };
+      // First argument is the source attribute path, all others are filter functions
+      if (typeof args[0] === 'string') {
+        var sourceAttributePath = args[0];
+        var filters = args.slice(1);
+        var value = sourceAttributePath.split('.').reduce(function(o, step) {return (o || {})[step];}, source);
```
Comment (etaoins) [inline index.js]: Nice :hammer:

File: README.md (hunk 1 above)
Comment (etaoins) [inline README.md]: I think you can leave it in with `let's use Cartography`

File: README.md (hunk 5 above)
Comment (xarvh) [inline README.md]: I don't even know how to explain this.
```
coffee> s = 'abcd'
'abcd'
coffee> typeof s
'string'
coffee> typeof new String s
'object'
```

Initially I was doing all convoluted check to allow String objects, but then the check was royally screwing up another part of my code in a horrible-to-fix way, so I decided that if you use Strings or Numbers or stuff you are a horrible human being and deserve to write your own custom filters anyway.

File: README.md (hunk 6 above)
Comment (xarvh) [inline README.md]: See above.
String literals will be rejected.

**Comment by etaoins** [comment]
LGTM :tropical_fish:

**Comment by etaoins** [comment]
:rocket: :earth_africa: :earth_americas: :earth_asia: :rocket:

**Pull request merged**